feat(batotov4): add new Bato.to V4 extension

LetianLi opened the pull request:
Adds a new extension for Bato.to V4.

This PR is independent of the one by @Sgwa in https://github.com/keiyoushi/extensions-source/pull/12320

I've tested popular, latest, search, and a lot of the filters. I've also updated the genre and language tags to what's listed on the site.  

Closes #12320, #12344

https://github.com/user-attachments/assets/51e7cc2b-e896-46f6-8cfb-0dcf7036251d

Checklist:
- [x] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [ ] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [x] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [x] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [ ] Have not changed source names
- [ ] Have explicitly kept the `id` if a source's name or language were changed
- [x] Have tested the modifications by compiling and running the extension through Android Studio
- [x] Have removed `web_hi_res_512.png` when adding a new extension

LetianLi committed:
Working search

LetianLi committed:
MVP: working details, chapter list, pages, webview.

LetianLi committed:
Rename to "Bato.to (V4)", replace last unstable q:key

LetianLi committed:
Fix chapter date upload

LetianLi committed:
Move search to GraphQL

LetianLi committed:
Move manga details to GraphQL

LetianLi committed:
Add upload status filter and fix manga webview

LetianLi commented:
The only issue at this moment is lack of signed in Utils and History filters. I don't quite understand what those even are.

LetianLi commented:
But everything else has been moved to GraphQL

LetianLi committed:
Add sser history/updates filters, and public comic list filter

LetianLi committed:
Rename back to "Bato.to V4"

aoba-seragakii commented:
Hello! I built the extension for testing. In the latest version, the status doesn't get fetched. Also, maybe it would be good adding the scanlator name as well, just like v2.

LetianLi committed:
Fix status parsing

LetianLi committed:
Add chapter's scanlator name

LetianLi commented:
Fixed the status parsing and added the scanlator's name.

However, I've mapped the status "pending" to `Smanga.UNKNOWN`, as I'm unsure what exactly that means. Let me know if it should map to ongoing instead.

LetianLi committed:
fix lint error

LetianLi commented:
I started working on this December 22nd after realizing that no one else had made an updated extension. 2 days later I had a working extension (mix of scraping and graphql), but realized that Sgwa had also made a PR (they happened to start work around the same time I did). 

Regardless, I created my PR, finished my graphql migration as I was already so close to finishing the migration from scraping to graphql, and installed it onto my phone so I can use it.

AwkwardPeak7 committed:
Refactor

AwkwardPeak7 committed:
Fix BatotoV4: make URLs in description clickable

AwkwardPeak7 committed:
Fix chapter number

AwkwardPeak7 committed:
Fix chapter number x2

AwkwardPeak7 committed:
nullable

AwkwardPeak7 committed:
fix filters not resetting

LetianLi commented:
The API doesn't support the user facing ascending/descending toggles that we currently have inside our SortFilter. 

Should we switch to a SelectFilter instead, or try to messily implement reverse pagination (sending first a request for how many pages are in the search, then doing a search starting from the last page, then reversing the result's order)?

LetianLi committed:
Refactor SortFilter to use SelectFilter
